src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManagerListener.cs

```diff
@@ -187,28 +178,29 @@ await _dispatcher
                 {ex.StackTrace}
                 """);
         }
+
+        return Task.CompletedTask;
     }
 
     private void Document_ChangedInEditor(object sender, EventArgs e)
     {
-        _ = Document_ChangedInEditorAsync(sender, CancellationToken.None);
+        Document_ChangedInEditorAsync(sender, CancellationToken.None).Forget();
     }
 
-    private async Task Document_ChangedInEditorAsync(object sender, CancellationToken cancellationToken)
+    private Task Document_ChangedInEditorAsync(object sender, CancellationToken cancellationToken)
     {
         try
         {
             // This event is called by the EditorDocumentManager, which runs on the UI thread.
             // However, due to accessing the project snapshot manager, we need to switch to
             // running on the project snapshot manager's specialized thread.
-            await _dispatcher
-                .RunAsync(() =>
-                    {
-                        var document = (EditorDocument)sender;
-                        _projectManager.DocumentChanged(document.ProjectKey, document.DocumentFilePath, document.EditorTextContainer!.CurrentText);
-                    },
-                    cancellationToken)
-                .ConfigureAwait(false);
+            return _projectManager.UpdateAsync(
+                updater =>
+                {
+                    var document = (EditorDocument)sender;
```

Comment:
Not going to comment everywhere, but you added an overload that took a `TState`, seems like `document` would be a good candidate?

Reply:
Agreed! I don't take the time to rewrite every usage in this pull request. Many of them were simple transformations. However, I can take a look through and tweak what I find.